lld/ELF/LinkerScript.cpp

```diff
@@ -717,14 +752,66 @@ void LinkerScript::processSectionCommands() {
       } else if (process(osec)) {
         osec->sectionIndex = i++;
       }
+    } else if (auto *sc = dyn_cast<SectionClassDesc>(base)) {
+      for (InputSectionDescription *isd : sc->sc.commands) {
+        isd->sectionBases =
+            computeInputSections(isd, ctx.inputSections, sc->sc);
+        for (InputSectionBase *s : isd->sectionBases) {
+          // Section classes with --enable-non-contiguous-regions may contain
+          // parented classes; spills for these are generated on reference.
+          if (!s->parent)
+            s->parent = &sc->sc;
+        }
+      }
+      sc->sc.assigned = true;
+    }
+  }
+
+  // Check that input sections cannot spill into or out of INSERT,
+  // since the semantics are nebulous. This is also true for OVERWRITE_SECTIONS,
+  // but no check is needed, since the order of processing ensures they cannot
+  // legally reference classes.
+  if (!potentialSpillLists.empty()) {
+    DenseSet<StringRef> insertNames;
+    for (InsertCommand &ic : insertCommands)
+      insertNames.insert(ic.names.begin(), ic.names.end());
+    for (SectionCommand *&base : sectionCommands) {
+      auto *osd = dyn_cast<OutputDesc>(base);
+      if (!osd)
+        continue;
+      OutputSection *os = &osd->osec;
+      if (!insertNames.contains(os->name))
+        continue;
+      for (SectionCommand *sc : os->commands) {
+        auto *isd = dyn_cast<InputSectionDescription>(sc);
+        if (!isd)
+          continue;
+        for (InputSectionBase *isec : isd->sectionBases)
+          if (isa<PotentialSpillSection>(isec) ||
+              potentialSpillLists.contains(isec))
+            errorOrWarn("section '" + isec->name +
+                        "' cannot spill from/to INSERT section '" + os->name +
+                        "'");
+      }
     }
+  }
 
   // If an OVERWRITE_SECTIONS specified output section is not in
   // sectionCommands, append it to the end. The section will be inserted by
   // orphan placement.
   for (OutputDesc *osd : overwriteSections)
     if (osd->osec.partition == 1 && osd->osec.sectionIndex == UINT32_MAX)
       sectionCommands.push_back(osd);
+
+  // Input sections cannot have a section class parent past this point; they
+  // must have been assigned to an output section.
+  for (const auto &[_, sc] : sectionClasses)
+    for (InputSectionDescription *isd : sc->sc.commands)
+      for (InputSectionBase *sec : isd->sectionBases)
+        if (sec->parent && isa<SectionClass>(sec->parent))
+          errorOrWarn("section '" + sec->name + "' assigned to class '" +
```

Comment:
The test shows one error/warning, but there are many. `.rodata.a` isn't very useful without an input file name.
```
ld.lld: error: section '.rodata.a' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.b' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.c' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.d' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.e' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.f' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.g' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.h' assigned to class 'a' but unreferenced by any output section
```

Perhaps we should report one single error: `section class 'a' is unreferenced`.

We don't need to mention matched input sections, since that's clear.

Reply:
Simplified the comment along these lines and made it only report once per class.